File: Python3/3100. Water Bottles II.py
class Solution:
    def maxBottlesDrunk(self, numBottles: int, numExchange: int) -> int:
        ans = empty = 0

        while empty + numBottles >= numExchange:
            ans += numBottles
            empty += numBottles
            if empty >= numExchange:
                numBottles = 1
                empty -= numExchange
                numExchange += 1

        return ans + numBottles

File: Java/3100. Water Bottles II.java
class Solution {
    public int maxBottlesDrunk(int numBottles, int numExchange) {
        int ans = numBottles, empty = numBottles;
        while (empty >= numExchange) {
            ans++;
            empty -= numExchange - 1;
            numExchange++;            
        }
        return ans;
    }
}